src/sage/rings/species.py

```diff
@@ -0,0 +1,2337 @@
+            True
+            sage: (G, {3: [1,2]}) in AtomicSpecies("X, Y")
+            False
+            sage: (G, {0: [1]}) in AtomicSpecies("X, Y")
+            False
+            sage: (G, {0: [1], 1: [2]}) in AtomicSpecies("X, Y")
+            False
+            sage: (0, {0: []}) in AtomicSpecies("X, Y")
+            False
+        """
+        if parent(x) == self:
+            return True
+        if isinstance(x, PermutationGroup_generic):
+            if self._arity == 1:
+                G = x
+                pi = {0: G.domain()}
+            else:
+                return False
+        else:
+            G, pi = x
+        if not isinstance(G, PermutationGroup_generic):
+            return False
+        if not set(pi.keys()).issubset(range(self._arity)):
+            return False
+        if (sum(len(p) for p in pi.values()) != len(G.domain())
+            or set(chain.from_iterable(pi.values())) != set(G.domain())):
+            return False
+        for orbit in G.orbits():
+            if not any(set(orbit).issubset(p) for p in pi.values()):
+                return False
+        return len(G.disjoint_direct_product_decomposition()) <= 1
+
+    def grading_set(self):
+        r"""
+        Return the set of non-negative integer vectors, whose length is
+        the arity of ``self``.
+
+        EXAMPLES::
+
+            sage: from sage.rings.species import AtomicSpecies
+            sage: AtomicSpecies(["X"]).grading_set()
+            Integer vectors of length 1
+        """
+        return IntegerVectors(length=self._arity)
+
+    def subset(self, size):
+        """
+        Return the set of atomic species with given total cardinality.
+
+        EXAMPLES::
+
+            sage: from sage.rings.species import AtomicSpecies
+            sage: A = AtomicSpecies(["X", "Y"])
+            sage: sorted(A.subset(3))
+            [E_3(X), C_3(X), E_3(Y), C_3(Y)]
+        """
+        result = Set()
+        for grade in IntegerVectors(size, length=self._arity):
+            result = result.union(self.graded_component(grade))
+        return result
+
+    def graded_component(self, mc):
+        """
+        Return the set of atomic species with given multicardinality.
+
+        EXAMPLES::
+
+            sage: from sage.rings.species import AtomicSpecies
+            sage: A = AtomicSpecies(["X", "Y"])
+            sage: len(A.graded_component([2, 3]))
+            1
+            sage: A.graded_component([2, 3])  # random
+            {{((2,3)(4,5), (1,2,3)): ({4, 5}, {1, 2, 3})}}
+        """
+        assert len(mc) == self._arity
+        S = SymmetricGroup(sum(mc)).young_subgroup(mc)
+        domain = S.domain()
+        pi = {i: domain[sum(mc[:i]): sum(mc[:i+1])] for i in range(len(mc))}
+        return Set([self(G, pi, check=False) for G in S.conjugacy_classes_subgroups()
+                    if len(G.disjoint_direct_product_decomposition()) <= 1])
+
+    @cached_method
+    def an_element(self):
```

Comment:
Rather than caching this, you can implement `def _an_element_(self):`, where the default `an_element()` uses that result and caches it.